### PR TITLE
iio-hyprland: init at 0-unstable-2024-07-24; nixos/iio-hyprland: init module

### PR DESCRIPTION
**Pull request by postgnostic**
## Description of changes

Listens to iio-sensor-proxy and automatically changes Hyprland output orientation.
Based on:
https://github.com/JeanSchoeller/iio-hyprland
WIth extra patch from:
https://github.com/short-circuit/iio-hyprland/commit/f13242bbdf4a0d062683bf23b2431bb72e2e6ab9

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/tools/wayland/hyprland-iio/default.nix
Comment (donovanglover) [inline pkgs/tools/wayland/hyprland-iio/default.nix:1]: Should be in [`pkgs/by-name`](https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name)

File: pkgs/tools/wayland/hyprland-iio/default.nix
Comment (donovanglover) [inline pkgs/tools/wayland/hyprland-iio/default.nix:1]: Remove the `...` and optionally use [`nixfmt-rfc-style`](https://search.nixos.org/packages?channel=unstable&show=nixfmt-rfc-style) to automatically format this file

File: pkgs/tools/wayland/hyprland-iio/default.nix
Comment (donovanglover) [inline pkgs/tools/wayland/hyprland-iio/default.nix:4]:
```suggestion
  pname = "iio-hyprland";
```

File: pkgs/tools/wayland/hyprland-iio/default.nix
Comment (donovanglover) [inline pkgs/tools/wayland/hyprland-iio/default.nix:5]:
```suggestion
  version = "0-unstable-2023-09-27";
```
Upstream hasn't tagged a release

File: pkgs/tools/wayland/hyprland-iio/default.nix
Comment (donovanglover) [inline pkgs/tools/wayland/hyprland-iio/default.nix:8]:
```suggestion
    owner = "JeanSchoeller";
```
Prefer pulling the main repo and using `fetchpatch` for your changes

File: pkgs/tools/wayland/hyprland-iio/default.nix
Comment (donovanglover) [inline pkgs/tools/wayland/hyprland-iio/default.nix:11]:
```suggestion
    hash = "0ahy72b66f3c0a0wydbcmn8zbfr335hv9jngw0zxh097z7h9sr2y";
```
Prefer `hash` over `sha256`

File: pkgs/tools/wayland/hyprland-iio/default.nix
Comment (donovanglover) [inline pkgs/tools/wayland/hyprland-iio/default.nix:14]: Build-time dependencies go in `nativeBuildInputs`

File: pkgs/tools/wayland/hyprland-iio/default.nix
Comment (donovanglover) [inline pkgs/tools/wayland/hyprland-iio/default.nix:18]:
```suggestion
    homepage = "https://github.com/JeanSchoeller/iio-hyprland";
```

File: pkgs/tools/wayland/hyprland-iio/default.nix
Comment (donovanglover) [inline pkgs/tools/wayland/hyprland-iio/default.nix:20]:
```suggestion
    maintainers = with maintainers; [ yusuf-duran ];
```
Makes it easier to add more maintainers in the future

**Review by donovanglover — COMMENTED** [review]
Welcome to nixpkgs!

File: nixos/modules/programs/iio-hyprland.nix
Comment (donovanglover) [inline nixos/modules/programs/iio-hyprland.nix:14]:
```suggestion
      enable = mkEnableOption
        (mdDoc "iio-hyprland and iio-sensor-proxy");
```

File: pkgs/by-name/hy/hyprland-iio/package.nix
Comment (donovanglover) [inline pkgs/by-name/hy/hyprland-iio/package.nix:25]:
```suggestion
    # Fix rotation of touch and tablet not working
    (fetchpatch {
```

**Review by donovanglover — COMMENTED** [review]
Should be good to go after adding `./programs/iio-hyprland.nix` to [`module-list.nix`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/module-list.nix).

File: nixos/modules/programs/iio-hyprland.nix
Comment (donovanglover) [inline nixos/modules/programs/iio-hyprland.nix:19]:
```suggestion
    hardware.sensor.iio.enable = mkDefault true;
```
Not needed since `with lib;`

File: nixos/modules/programs/iio-hyprland.nix
Comment (eclairevoyant) [inline nixos/modules/programs/iio-hyprland.nix:3]:
```suggestion
```
bad practice from older nixpkgs code

File: nixos/modules/programs/iio-hyprland.nix
Comment (eclairevoyant) [reply nixos/modules/programs/iio-hyprland.nix:19]: `with lib;` should be discouraged instead

File: nixos/modules/programs/iio-hyprland.nix
Comment (eclairevoyant) [reply nixos/modules/programs/iio-hyprland.nix:14]: `mdDoc` does nothing, I'd simply suggest

```suggestion
      enable = mkEnableOption "iio-hyprland and iio-sensor-proxy";      
```

Comment (Yusuf-Duran) [reply nixos/modules/programs/iio-hyprland.nix]: I've done this without the mdDoc for now, if markdown is to be added I will add it to my module accordingly.

File: nixos/modules/programs/iio-hyprland.nix
Comment (Yusuf-Duran) [reply nixos/modules/programs/iio-hyprland.nix:19]: I have removed the `with lib;` instead and (hopefully) added `lib.` wherever it is needed instead.

**Review by donovanglover — APPROVED** [review]
LGTM

**Comment by Yusuf-Duran** [comment]
@eclairevoyant could you please review the changes?

File: pkgs/by-name/hy/hyprland-iio/package.nix
Comment (eclairevoyant) [inline pkgs/by-name/hy/hyprland-iio/package.nix:1]: Shouldn't this be packaged under `iio-hyprland` not `hyprland-iio`? Everything else seems to be named the former.

File: pkgs/by-name/hy/hyprland-iio/package.nix
Comment (eclairevoyant) [inline pkgs/by-name/hy/hyprland-iio/package.nix:30]:
```suggestion
    (fetchpatch2 {
      name = "keyword.patch";
      url = "https://github.com/JeanSchoeller/iio-hyprland/pull/14/commits/f13242bbdf4a0d062683bf23b2431bb72e2e6ab9.diff";
      hash = "sha256-RRjyVTPJj3+7q1eqqQXnH6izCa4fLbYXV6oWJqHuQFs=";
    })
```

As is, it seemed like the commit was from the main upstream, not a fork. This should make it clearer that it comes from a PR.
And I believe `fetchpatch2` is currently preferred.

**Review by eclairevoyant — DISMISSED** [review]
I don't have a hyprland setup to test this on, but I noticed a couple more points:

File: pkgs/by-name/hy/hyprland-iio/package.nix
Comment (JohnRTitor) [reply pkgs/by-name/hy/hyprland-iio/package.nix:30]: https://github.com/JeanSchoeller/iio-hyprland/pull/14 got merged a while ago, so let's move to the latest unstable commit.

File: pkgs/by-name/hy/hyprland-iio/package.nix
Comment (JohnRTitor) [reply pkgs/by-name/hy/hyprland-iio/package.nix:1]: Upstream repo is `iio-hyprland`, so it makes sense to use that. Done.

**Review by JohnRTitor — APPROVED** [review]
Looks good to me

**Comment by sashapop10** [comment]
Shouldn't it be run as a service? I didn't test it, but with quick look it looks broken. Enable option doesn't execute package it self on boot. It looks like user need to do it manualy now...

```nix
{ 
  systemd.services.iio-hyprland = {
    enable = true;
    name = "iio-hyprland";
    description = "Enables iio-hyprland";
    wantedBy = [ "multi-user.target" ];
    serviceConfig = { 
      Type = "oneshot";
      Restart = "no";
      ExecStart = "${pkgs.iio-hyprland}/bin/iio-hyprland";
    };
  };
}
```

**Pull request merged**
